core/src/com/projectkorra/projectkorra/object/HorizontalVelocityTracker.java

```diff
@@ -86,33 +92,36 @@ public void update() {
 
 		final Vector diff = this.thisVelocity.subtract(this.lastVelocity);
 
-		final List<Block> blocks = GeneralMethods.getBlocksAroundPoint(this.entity.getLocation(), 1.5);
+		if (this.thisVelocity.length() < this.lastVelocity.length()) {
+			if ((diff.getX() > 0 || diff.getX() < 0) || (diff.getZ() > 0 || diff.getZ() < 0)) {
+				this.impactLocation = this.entity.getLocation();
 
-		for (final Block b : blocks) {
-			if (ElementalAbility.isWater(b)) {
-				this.remove();
-				return;
+				if (didHitWall((LivingEntity) this.entity)) {
+					this.hasBeenDamaged = true;
+					ProjectKorra.plugin.getServer().getPluginManager().callEvent(new HorizontalVelocityChangeEvent(this.entity, this.instigator, this.lastVelocity, this.thisVelocity, diff, this.launchLocation, this.impactLocation, this.abil));
+					this.remove();
+					return;
+				}
 			}
 		}
+	}
 
-		if (this.thisVelocity.length() < this.lastVelocity.length()) {
-			if ((diff.getX() > 1 || diff.getX() < -1) || (diff.getZ() > 1 || diff.getZ() < -1)) {
-				this.impactLocation = this.entity.getLocation();
-				for (final Block b : blocks) {
-					if (b.getType() == Material.BARRIER && !this.barrier) {
-						return;
-					}
-					if (GeneralMethods.isSolid(b) && (this.entity.getLocation().getBlock().getRelative(BlockFace.EAST, 1).equals(b) || this.entity.getLocation().getBlock().getRelative(BlockFace.NORTH, 1).equals(b) || this.entity.getLocation().getBlock().getRelative(BlockFace.WEST, 1).equals(b) || this.entity.getLocation().getBlock().getRelative(BlockFace.SOUTH, 1).equals(b))) {
-						if (!ElementalAbility.isTransparent(this.instigator, b)) {
-							this.hasBeenDamaged = true;
-							ProjectKorra.plugin.getServer().getPluginManager().callEvent(new HorizontalVelocityChangeEvent(this.entity, this.instigator, this.lastVelocity, this.thisVelocity, diff, this.launchLocation, this.impactLocation, this.abil));
-							this.remove();
-							return;
-						}
-					}
+	private boolean didHitWall(LivingEntity entity) {
+		BlockFace[] faces = { BlockFace.EAST, BlockFace.NORTH, BlockFace.WEST, BlockFace.SOUTH, BlockFace.NORTH_EAST, BlockFace.NORTH_WEST, BlockFace.SOUTH_EAST, BlockFace.SOUTH_WEST };
```

Comment:
Do you still want to prune this array down to only the BlockFaces in the opposite direction of the entity's velocity?

Reply:
I do, but this still works fine without any complaints from the community (as far as I remember from their feedback). But I remember @Dreig-Michihi had a better fix for addressing this issue? Wanted to see if they wanted to do a PR so I could close this and theirs could be implemented instead, but I guess not. Regardless, I think this should be fine.

Reply:
once this is merged I can change it, there are some other things I want to cleanup in this class at the same time lol

Reply:
We'll be saving this fix for a future update lol. The problems for this issue run a bit deeper than this naive fix according to @StrangeOne101 and @Dreig-Michihi.